core/trino-spi/src/test/java/io/trino/spi/TestSingleVersionBackwardCompatibility.java

```diff
@@ -0,0 +1,123 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import io.trino.spi.connector.ConnectorContext;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.Sets.difference;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+import static java.lang.ClassLoader.getSystemClassLoader;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSingleVersionBackwardCompatibility
+{
+    private static final Map<String, Set<String>> BACKWARD_INCOMPATIBLE_CHANGES = ImmutableMap.<String, Set<String>>builder()
+            // When updating this map, please try to remove backward incompatible changes for old versions.
+            // Also consider mentioning backward incompatible changes in release notes.
+            .put("375", ImmutableSet.of(
+                    "Constructor: public io.trino.spi.connector.TableProcedureExecutionMode(boolean,boolean)",
+                    "Constructor: public io.trino.spi.eventlistener.QueryStatistics(java.time.Duration,java.time.Duration,java.time.Duration,java.time.Duration,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,java.util.Optional<java.time.Duration>,long,long,long,long,long,long,long,long,long,long,long,long,long,long,long,double,double,java.util.List<io.trino.spi.eventlistener.StageGcStatistics>,int,boolean,java.util.List<io.trino.spi.eventlistener.StageCpuDistribution>,java.util.List<java.lang.String>,java.util.Optional<java.lang.String>)"))
+            .buildOrThrow();
+
+    @Test
+    public void testSpiSingleVersionBackwardCompatibility()
+            throws Exception
+    {
+        Set<String> backwardIncompatibleChanges = BACKWARD_INCOMPATIBLE_CHANGES.getOrDefault(getPreviousSpiVersion(), ImmutableSet.of());
+        assertThat(getCurrentSpi()).containsAll(difference(getPreviousSpi(), backwardIncompatibleChanges));
+    }
+
+    private static String getPreviousSpiVersion()
+    {
+        int currentVersion = Integer.parseInt(new DummyConnectorContext().getSpiVersion().replace("-SNAPSHOT", ""));
+        return Integer.toString(currentVersion - 1);
```

Comment:
This is generally the case, but might not always be. For instance, if something happens during the release process, we might need to "burn" a release (e.g., https://trino.io/docs/current/release/release-349.html)

Reply:
we will have to change this code line when doing that. perhaps add an `if`.

Reply:
note also that we would have same problem with JDBC and server compatibility testing. 
A rare case of a burned release would need to be handled explicitly.